sound/soc/sof/mediatek/adsp_pcm.c

```diff
@@ -0,0 +1,224 @@
+			struct adsp_mem *mem_info)
+{
+	int ret;
+
+	mem_pool->dram_pool = gen_pool_create(SOF_MIN_ADSP_SHIFT, -1);
+	if (!mem_pool->dram_pool)
+		return -ENOMEM;
+
+	if (!mem_info->va_addr || !mem_info->size) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	ret = gen_pool_add_virt(mem_pool->dram_pool, mem_info->va_addr,
+				mem_info->phy_addr, mem_info->size, -1);
+	if (ret)
+		goto out;
+
+	mem_pool->dram = mem_info;
+
+	return 0;
+
+out:
+	gen_pool_destroy(mem_pool->dram_pool);
+	mem_pool->dram_pool = NULL;
+	return ret;
+}
+EXPORT_SYMBOL_NS(adsp_genpool_create, SND_SOC_SOF_MTK_COMMON);
+
+int adsp_genpool_destroy(struct adsp_mem_pool *mem_pool)
+{
+	gen_pool_destroy(mem_pool->dram_pool);
+	mem_pool->dram_pool = NULL;
+	mem_pool->dram = NULL;
+	return 0;
+}
+EXPORT_SYMBOL_NS(adsp_genpool_destroy, SND_SOC_SOF_MTK_COMMON);
+
+int adsp_genpool_alloc(struct adsp_mem_pool *mem_pool, unsigned char **vaddr,
+		       dma_addr_t *paddr, unsigned int size)
+{
+	if (!mem_pool || !mem_pool->dram_pool)
+		return -EINVAL;
+
+	*vaddr = (unsigned char *)gen_pool_dma_alloc(mem_pool->dram_pool,
+						     size, paddr);
+	if (!*vaddr)
+		return -ENOMEM;
+
+	return 0;
+}
+
+int adsp_genpool_alloc_align(struct adsp_mem_pool *mem_pool,
+			     unsigned char **vaddr, dma_addr_t *paddr,
+			     unsigned int size, int align)
+{
+	if (!mem_pool || !mem_pool->dram_pool)
+		return -EINVAL;
+
+	*vaddr = (unsigned char *)gen_pool_dma_alloc_align(mem_pool->dram_pool,
+							   size, paddr, align);
+	if (!*vaddr)
+		return -ENOMEM;
+
+	return 0;
+}
+
+int adsp_genpool_free(struct adsp_mem_pool *mem_pool,
+		      unsigned char *vaddr, size_t size)
+{
+	if (!mem_pool || !mem_pool->dram_pool)
+		return -EINVAL;
+
+	if (*vaddr)
+		gen_pool_free(mem_pool->dram_pool, (unsigned long)vaddr, size);
+
+	return 0;
+}
+
+int mtk_adsp_pcm_hw_params(struct snd_sof_dev *sdev,
+			   struct snd_pcm_substream *substream,
+			   struct snd_pcm_hw_params *params,
+			   struct sof_ipc_stream_params *ipc_params)
+{
+	struct snd_pcm_runtime *runtime = substream->runtime;
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_component *scomp = sdev->component;
+	struct device *dev = sdev->dev;
+	struct snd_sof_pcm *spcm;
+	struct snd_dma_buffer *dmab;
+	struct snd_dma_buffer *pg_table;
+	struct snd_dma_buffer *pg_table_backup;
+	struct adsp_priv *priv = sdev->pdata->hw_pdata;
+	struct adsp_mem_pool *mem_pool = &priv->mem_pool;
+	size_t dma_bytes;
+	int ret;
+
+	dma_bytes = runtime->dma_bytes;
+
+	spcm = snd_sof_find_spcm_dai(scomp, rtd);
+	if (!spcm) {
+		dev_err(sdev->dev, "%s : can't find PCM with DAI ID %d\n",
+			__func__, rtd->dai_link->id);
+		return -EINVAL;
+	}
+
+	pg_table = &spcm->stream[substream->stream].page_table;
+	if (runtime->buffer_changed && substream->managed_buffer_alloc) {
+		/* free dma pages allocated by common layer
+		 * if page table private data is not NULL,
+		 * this hw_params may be trigged by PCM xrun
```

Comment:
xrun will result in .prepare being called, not hw_params. There must be a confusion here?